Report Benchmarking Method Added

sadrasabouri opened the pull request:
#### Reference Issues/PRs
#604 and #605 

#### What does this implement/fix? Explain your changes.
This PR provides an API for reporting the benchmark of creation of confusion matrices.
By adding `report_benchmark` function to `generate_random_data` module, users/developers can assess the time it takes for the

- Confusion matrix creation
- Calculating the class metrics
- Calculating the overall metrics

in addition to the total time for a variety range of population and class sizes under different scenarios regarding to the data imbalance.

#### Any other comments?

sepandhaghighi commented on pycm/generate_random_data.py:
```diff
@@ -139,3 +143,36 @@ def generate_confusion_matrix_with_scenario(
+
+def run_report_benchmark(seed=None):
+    """
+    Benchmark the generation of a confusion matrix and print the report.
```
I believe it should be `some confusion matrices`

sadrasabouri commented:
@sepandhaghighi @alirezazolanvari
Guys I identified a seed problem and fixed that and added some documentation for what you suggested.
Let me know what you think.

sepandhaghighi commented on pycm/generate_random_data.py:
```diff
@@ -69,16 +74,20 @@ def _calculate_class_counts(class_percentages, total_population):
+    :type seed: int or None
     :return: confusion matrix as a dictionary
     """
+    if seed is not None:
```
I recommend eliminating this condition and instead calling `np.random.seed(seed)` directly.

sepandhaghighi commented on pycm/generate_random_data.py:
```diff
@@ -138,4 +151,35 @@ def generate_confusion_matrix_with_scenario(
                 [sen.value for sen in ClassDistributionScenario]))
     class_percentages = _generate_class_percentages(num_classes, scenario)
-    return generate_confusion_matrix(class_percentages, total_population)
+    return generate_confusion_matrix(class_percentages, total_population, seed=seed)
```
`generate_confusion_matrix(class_percentages=class_percentages, total_population=total_population, seed=seed)`

sepandhaghighi commented on pycm/params.py:
```diff
@@ -879,6 +879,19 @@
+BENCHMARK_REPORT_TEMPLATE = """Number of classes: {num_classes}
+Population size: {total_population}
+Class distribution scenario: {scenario}
+Timing:
```
```
Timing:
    + Matrix creation: {timing_matrix_creation:.5f}s
    + Class statistics: {timing_class_statistics:.5f}s
    + Overall statistics: {timing_overall_statistics:.5f}s
    + Total: {timing_total:.5f}s
======================================================
```

sepandhaghighi commented:
@alirezazolanvari This is just a friendly reminder ⏲️

sadrasabouri commented:
Applied the changes. It's ready for re-review now @sepandhaghighi @alirezazolanvari

alirezazolanvari commented on Document/Document.ipynb:
```diff
@@ -14761,6 +14762,71 @@
+    "\n",
+    "You can generate random confusion matrices based on these scenarios using the following code:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from pycm import generate_confusion_matrix_with_scenario, ClassDistributionScenario\n",
+    "\n",
+    "for scenario in ClassDistributionScenario:\n",
+    "    rand_cm_dict = generate_confusion_matrix_with_scenario(\n",
+    "        num_classes=3,\n",
+    "        total_population=1000,\n",
+    "        scenario=scenario,\n",
+    "        seed=0)\n",
+    "    rand_cm = ConfusionMatrix(matrix=rand_cm_dict)\n",
+    "    print(f\"Confusion matrix generated with scenario {scenario.name}\")\n",
+    "    rand_cm.print_matrix()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Using various instances of these scenarios, you can benchmark `pycm`'s runtime performance with the `run_report_benchmark` function. This  evaluates performance over a predefined, diverse set of class counts and sample sizes, helping assessing how `pycm` scales under different data conditions."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from pycm import run_report_benchmark\n",
+    "run_report_benchmark(seed=0)"
+   ]
+  },
```
```suggestion
 {
   "metadata": {},
   "cell_type": "markdown",
   "source": "PyCM offers a dedicated benchmarking method for performance analysis across different hardware. Using this utility, you can benchmark `pycm`'s runtime performance with the `run_report_benchmark` function. This allows users to measure the time required for various internal operations, such as generating a confusion matrix and calculating metrics, on their own hardware, using predefined test cases."
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from pycm import run_report_benchmark\n",
    "run_report_benchmark(seed=0)"
   ]
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    "The predefined test cases are offered as a set of predefined data distribution scenarios through the `ClassDistributionScenario` class. With this you can simulate class imbalance in various ways. The available options include:\n",
    "\n",
    "+ `ClassDistributionScenario.UNIFORM`: All classes have equal frequency (balanced distribution).  \n",
    "+ `ClassDistributionScenario.MAJORITY_CLASS`: One class appears 5 times more frequently than the others, which have equal frequency.  \n",
    "+ `ClassDistributionScenario.MINORITY_CLASS`: One class appears 5 times less frequently than the others, which have equal frequency.\n",
    "\n",
    "You can generate random confusion matrices based on these scenarios using the following code:"
   ]
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": [
    "from pycm import generate_confusion_matrix_with_scenario, ClassDistributionScenario\n",
    "\n",
    "for scenario in ClassDistributionScenario:\n",
    "    rand_cm_dict = generate_confusion_matrix_with_scenario(\n",
    "        num_classes=3,\n",
    "        total_population=1000,\n",
    "        scenario=scenario,\n",
    "        seed=0)\n",
    "    rand_cm = ConfusionMatrix(matrix=rand_cm_dict)\n",
    "    print(f\"Confusion matrix generated with scenario {scenario.name}\")\n",
    "    rand_cm.print_matrix()"
   ]
  },
```

sepandhaghighi commented on pycm/generate_random_data.py:
```diff
@@ -1,7 +1,13 @@
 # -*- coding: utf-8 -*-
 """This file contains a function to generate a random confusion matrix."""
+from unicodedata import digit
```
Is this an unused import?
